README.md
# BITS

* [Quickstart](#quickstart)
  * [Application Arguments](#application-arguments)
  * [Environment Variables](#environment-variables)
* [Modules](#modules)
  * [Available Modules](#available-modules)
  * [Module Basics](#module-basics)
    * [Scopes](#scopes)
    * [Load Policy](#load-policy)
    * [module.json](#modulejson)
    * [index.js](#indexjs)
* [global](#global)
* [APIs](#apis)
* [Helpful Links](#helpful-links)

BITS Integrated Technology System (BITS) is designed for the rapid development of modules that share base software. The goal is to reuse common components in multiple projects and harden these components through extensive use.

BITS can run on enterprise or embedded applications and can help jumpstart a project with initial functionality.

A basic BITS implementation provides the following initial capabilities:

* Module Management
* Inter-module Messaging ([Message Center](https://github.com/LGSInnovations/node-bits-message-center))
* Authorization Scopes

See [Available Modules](#available-modules) for additional capabilities.

## Quickstart

Built the NodeJs framework, BITS requires [NodeJS v12](https://nodejs.org/en/download/).

``` bash
# Download BITS
git clone https://github.com/LGSInnovations/bits.git

# Build
cd bits
npm run build

# Start
npm run dev
```

### Application Arguments

Arg | Description | Default | Example
--- | --- | --- | ---
*d* | Specifiy a path to the data directory. | &lt;projectRoot&gt;/data | `node app.js -d /full/path/to/dir`
*m* | Specifiy a path to the modules directory. | &lt;dataDir&gt;/base/modules/modules | `node app.js -m /full/path/to/modules`
*o* | Specifiy a path to the log file. | *null* | `node app.js -o /full/path/to/logs`
*v* | Set verbose logging. | *false* | `node app.js -v`

### Environment Variables

Env | Description | Default | Example
--- | --- | --- | ---
*HTTP_PORT* | Specify a port for the HTTP server. | *80* | `HTTP_PORT=9000 node app.js`
*HTTPS_PORT* | Specify a port for the HTTPS server. | *443* | `HTTP_PORT=9001 node app.js`

**Note** BITS does not set up a web server, the [BITS Web Server](https://github.com/LGSInnovations/bits-webserver) module does, however, the environment variable is set through the initial base process.

## Modules

BITS is a framework built around modules. The base BITS framework is made useful with the creation and addition of module functionality. A module should provide a concrete feature to the system.

### Available Modules

* [Authentication](https://github.com/LGSInnovations/bits-authentication)
* [Auth Bearer Strategy](https://github.com/LGSInnovations/bits-auth-bearer)
* [Credentials](https://github.com/LGSInnovations/bits-credentials)
* [Crypto](https://github.com/LGSInnovations/bits-crypto)
* [Google Polymer](https://github.com/LGSInnovations/bits-ui-polymer)
* [Logging Exports](https://github.com/LGSInnovations/bits-logging)
* [MongoDB](https://github.com/LGSInnovations/bits-mongodb)
* [Network Manager](https://github.com/LGSInnovations/bits-network-manager)
* [Node IPC](https://github.com/LGSInnovations/bits-node-ipc)
* [Notifications](https://github.com/LGSInnovations/bits-notifications)
* [Optimized Module Groupings](https://github.com/LGSInnovations/bits-omg)
* [Single Page Application](https://github.com/LGSInnovations/bits-ui-spa)
* [Web Server](https://github.com/LGSInnovations/bits-webserver)
* [Users](https://github.com/LGSInnovations/bits-users)

### Module Basics

#### Scopes

BITS uses an attribute based architecture for access control. Each requester is assigned a set of scopes. The requester is then only able to make requests or receive events that are tagged with the same scope as the request/event is tagged with. Normally, when a request or event listener is added to the message center, the author specifies the scopes attribute in the second parameter.

Notify BITS of your scopes through the module.json.

#### Load Policy

BITS will, by default, attempt to load modules exactly once. If they succeed, the process will continue running until told otherwise. A failure will result in the module not being loaded. Modules may, however, override this behavior by specifying a load policy in the `module.json`.

The load policy is configured by adding a JSON object to the file with the key `load`. This object may contain the following key/value pairs:

* restartPolicy
  * never (DEFAULT)
    * BITS will never re-attempt loading the module
  * oneshot
    * BITS will load the module, but the process will be stopped when it runs to completion
    * If the module load fails, BITS will retry, as specified by the `retries` value
  * on-failure
    * BITS will load the module as usual
    * If the module load fails, BITS will retry, as specified by the `retries` value
* retries
  * The number of times to retry loading
    * Negative numbers will retry indefinitely
    * Default is 1


#### module.json

The module.json defines metadata that will be used by BITS to load your module. Additionally, that metadata will be available on your module's process as `process.env.mod`.

Key | Type | Description
--- | --- | ---
*name* | *String* | A unique name for your module.
*version* | *String* | Semantic Version of your module. Our build process automatically fills in the version based off the specified tag in the repository. You will commonly see this blank, but it should have a value in a production system.
*scopes* | *Array* | Array of scope definitions.
*load* | *Object* | Load Policy.
*dependencies* | *Object* | A list of dependencies and their versions. Note, use the name of the module from its module.json.

```javascript
// module.json
{
  "name": "my-module",
  "version": "1.2.3",
  "scopes": [{
    "name": "my-module-scopes",
    "displayName": "My Module Scopes"
  },...],
  "dependencies": {
    "bits-base": "^6.0.0",
    "dependency-module": "^1.0.0",
    ...
  }
}
```

```javascript
// module.json with load policy
{
  "name": "my-module",
  "version": "1.2.3",
  "scopes": [{
    "name": "my-module-scopes",
    "displayName": "My Module Scopes"
  },...],
  "load": {
    "restartPolicy": "oneshot",
    "retries": 2
  },
  "dependencies": {
    "bits-base": "^6.0.0",
    "dependency-module": "^1.0.0",
    ...
  }
}
```

#### index.js

An `index.js` needs to be specified to run module code. The `index.js` should export two functions, load and unload. These methods are passed the `messageCenter` for inter-process communication.

```javascript
// index.js
(() => {
  'use strict';

  class MyModule {

    load(messageCenter) {
      return Promise.resolve();
    }

    unload(messageCenter) {
      return Promise.resolve();
    }
  }

  module.exports = new MyModule();
})();
```

When using `oneshot` as the `restartPolicy`, the `load`/`unload` functions should be `static`, and the `index.js` itself should be stateless.

```javascript
// index.js - oneshot reload policy
(() => {
  'use strict';

  class MyModule {
    static load(messageCenter) {
      return Promise.resolve()
      .then(() => {
        // stateless activities
      });
    }

    static unload(messageCenter) {
      return Promise.resolve()
      .then(() => {
        // undo everything, as needed.
      });
    }
  }

  module.exports = MyModule;
})();
```

## global

The `global` object is added to and replicated across all modules. While it makes sense in some scenarios to add to the `global` it is best practice to use the message center to transfer data and make files available for loading. Therefore, the `global` is used sparingly.

**`global.paths`** (*Object*) - file system paths.

Key | Type | Description
--- | --- | ---
*base* | *String* | Absolute path to BITS on the file system.
*data* | *String* | Absolute path to the data directory on the file system.
*modules* | *String* | Absolute path to the modules directory on the file system.
*moduleDataDir* | *String* | Absolute path to your module's directory on the file system.

**`global.bitsId`** (*String*) - the BITS ID of the system. See [bits-id.js](./lib/base/bits-id.js).

**`global.helper`** (*LazyRequireLoader*) - add/remove functionality for key/value. Key is the name of the API being added, and value is the file path.

```javascript
// to add to the helper
global.helper.add('MyClassName', '/path/to/my/class');

// to use the helper in another module
const MyClassName = global.helper.MyClassName;
```

See [lazy-require-loader.js](./lib/utils/lazy-require-loader.js).

## APIs

See the following APIs:

* [Dispatcher](./lib/dispatcher/api.js)
* [Helper](./lib/helper/api.js) - **Note** `add` and `remove` already exist on the `global.helper`, see [global](#global).
* [Modules](./lib/modules/api.js)
* [Scopes](./lib/scopes/api.js)

## Helpful Links

There have been many NPM packages published to remove duplicate code, normalize module structure, and apply tried and true design patterns to a module architecture. Below are regularly maintained BITS repositories.

**Style/Formatting**

* [BITS Style Guide](https://lgsinnovations.github.io/guidelines-bits/styleguide/javascript.html)
* [BITS ESlint Config](https://github.com/LGSInnovations/eslint-config-bits)
* [BITS ESlint](https://github.com/LGSInnovations/node-bits-eslint)

**Helpers**

* [Core](https://github.com/LGSInnovations/node-bits-core)
* [Memory CRUD](https://github.com/LGSInnovations/node-bits-memory-crud)
* [PouchDB CRUD](https://github.com/LGSInnovations/node-bits-pouch-crud)
* [MongoDB CRUD](https://github.com/LGSInnovations/node-bits-mongo-crud)
* [Memory Key/Value](https://github.com/LGSInnovations/node-bits-memory-kv)
* [LevelDB Key/Value](https://github.com/LGSInnovations/node-bits-level-kv)

**Utilities**

* [Message Center](https://github.com/LGSInnovations/node-bits-message-center)
* [Socket.io Client](https://github.com/LGSInnovations/node-bits-socket.io-client)
* [Logger](https://github.com/LGSInnovations/node-bits-logger)
* [Auto Discovery](https://github.com/LGSInnovations/node-bits-auto-discovery)
* [Daemon](https://github.com/LGSInnovations/node-bits-daemon)
* [File System](https://github.com/LGSInnovations/node-bits-fs)
* [Scopes](https://github.com/LGSInnovations/node-bits-scopes-util)
* [Module Lazy Load](https://github.com/LGSInnovations/node-bits-lazy-load)
* [Crypto](https://github.com/LGSInnovations/node-bits-crypto)


CHANGELOG.md
<a name="5.0.1"></a>
## 5.0.1 (2019-10-18)

---

<a name="5.0.0"></a>
# 5.0.0 (2019-08-14)


### Bug Fixes

* ensure <data-dir>/tmp exists   (110b6d5), closes PNP-1854
* prevent fatal error on load if socket is not fully initialized   (20180e1), closes pnp-2194


### Features

* add in a global path for base so all modules can get running base dir   (dd0a06e)
* add support to export from journalctl   (c1ccb69), closes PNP-1252
* remove OMG subsystem   (805d2c1)


### BREAKING CHANGES

* removes OMG subsystem and ability to upgrade base. See the bits-omg module for continued support of system upgrade.

---

<a name="4.1.0"></a>
# 4.1.0 (2019-06-04)


### Bug Fixes

* **daemon:** emit the 'exit' event   (5574176), closes PNP-1745


### Features

* adds origin to message center metadata   (e06bd86), closes PNP-1705

---

<a name="4.0.0"></a>
# 4.0.0 (2019-05-24)


### Bug Fixes

* Allow routes to be added when classes extend base crud router   (4f3f1a8), closes BITS-255
* bind to module to prevent deletion race condition   (f9d93b3), closes PNP-1573
* changes to readstream API   (40e48ba)
* correctly compare scopes with development scope   (85ce91f), closes PNP-1389
* default undefined arguments in crud-api to be empty objects   (c914a6b), closes PNP-118
* delete the correct module on exit   (a488c90), closes PNP-1322 PNP-1407
* do not overwrite load error if one exists   (332cfce), closes PNP-1587
* fix upgrade-base for node10   (e278ef7)
* log messages fail to be properly parsed   (3721926), closes PNP-1605
* remove additional endpoints for delete   (eda498f), closes PNP-1356
* send signal to process.kill() to determine how process was killed   (31f3ceb), closes PNP-1429
* **logger:** format the logger output to the file for human-readability   (07674e3), closes PNP-1532
* winston splat handler   (3cf9438)
* winston work-around for logging with 2 args   (dd4d100)


### Features

* add API to retrieve display name for a module   (1d5b631), closes PNP-1388
* adds root router resource listener to base server   (f3336c5), closes PNP-1485
* event listeners on workers   (db3fd5f), closes PNP-1573 PNP-1562
* remove endpoint in helper API   (8db0cbd), closes PNP-1293
* **certificate:** add trusted ca and api to get trusted ca   (097f61c), closes GARB-483
* **certificate/key:** add cert & key classes to utils   (99ff86a), closes BITS-248 GARB-534
* **load-policy:** add load policy to allow modules to specify how the process should be maintained/   (96e8c36), closes PNP-1322
* remove public scope   (c8a0f17)
* removes web based user interface   (aff7617), closes PNP-1268
* update node dependencies   (688472e), closes BITS-197
* winston 3.x.x upgrade   (4479de2), closes BITS-198


### BREAKING CHANGES

* remove "public" scope; the motivation being that everyone gets
	the public scope, so why bother?

	Before:
	scopes: ['public']

	After:
	scopes: []
* newer NPM dependencies will no longer work in node 6
* removes all Polymer elements and API's for pages, gallery,
toolbar, and widgets

---

<a name="3.5.0"></a>
# 3.5.0 (2019-04-10)


### Bug Fixes

* allow bits to run from any directory   (44f9a0f)
* allow upgrade scripts to run without progress server   (c1e3e9f), closes BITS-157
* assign value to the correct property   (ba8c3f9), closes NMON-432
* **activities:** require user interaction to dismiss notifications   (c94c410), closes PNP-1271
* **OmgApi:** resolve add() once complete   (fa7a243), closes EDG-68
* **routers:** fix body parsing for all CRUD routers   (4c1da42), closes PNP-113
* **upgrade-server:** Upgrade server should not wait for server connection   (cd08d3e), closes PNP-587
* check error against "not found" regex   (aacf8f8), closes GARB-605
* create a module dir in tmp for unit tests   (36f9595), closes PNP-822
* ensure childProcess object exists before calling kill on it   (5c03c83)
* PouchDB _id initialize   (ee03444), closes PNP-903
* toolbar items observer listens on length changes   (85bbcaa), closes PNP-885


### Features

* **display-name:** add behavior to pull display name from module.json   (6eea18a), closes PNP-580
* **OmgsApi:** expose delete   (9561384), closes EDG-67
* Add a field to the child process utility to pass stdin to the process   (cd2be64), closes PNP-1116
* Add a symlink wrapper to the UtilFs utility   (ba8285f), closes PNP-1116
* expose BaseOmgsApi   (6faad49), closes EDG-64
* make modules dir a configurable option on the command line   (56082d6), closes BITS-163

---

<a name="3.5.0"></a>
# [3.5.0] (2019-04-05)


### Bug Fixes

* allow bits to run from any directory (44f9a0f)
* allow upgrade scripts to run without progress server (c1e3e9f), closes BITS-157
* assign value to the correct property (ba8c3f9), closes NMON-432
* **OmgApi:** resolve add() once complete (fa7a243), closes EDG-68
* **routers:** fix body parsing for all CRUD routers (4c1da42), closes PNP-113
* **upgrade-server:** Upgrade server should not wait for server connection (cd08d3e), closes PNP-587
* check error against "not found" regex (aacf8f8), closes GARB-605
* create a module dir in tmp for unit tests (36f9595), closes PNP-822
* ensure childProcess object exists before calling kill on it (5c03c83)
* PouchDB _id initialize (ee03444), closes PNP-903
* toolbar items observer listens on length changes (85bbcaa), closes PNP-885


### Features

* **display-name:** add behavior to pull display name from module.json (6eea18a), closes PNP-580
* **OmgsApi:** expose delete (9561384), closes EDG-67
* Add a field to the child process utility to pass stdin to the process (cd2be64), closes PNP-1116
* Add a symlink wrapper to the UtilFs utility (ba8285f), closes PNP-1116
* expose BaseOmgsApi (6faad49), closes EDG-64
* make modules dir a configurable option on the command line (56082d6), closes BITS-163

---

<a name="3.4.0"></a>
# 3.4.0 (2019-01-21)


### Bug Fixes

* **auth:** retrieve a user object for any registered authentication strategy   (58e8042), closes BITS-246

lib/base-server.js
/**
Copyright 2017 LGS Innovations

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/
(() => {
  'use strict';

  const HTTP_PORT = process.env.HTTP_PORT || 80;
  const HTTPS_PORT = process.env.HTTPS_PORT || 443;

  const os = require('os');
  const path = require('path');
  const http = require('http');
  const https = require('https');
  const express = require('express');
  const EventEmitter = require('events');
  const helmet = require('helmet');
  const compression = require('compression');
  const UtilFs = require('@lgslabs/bits-fs');
  const UtilPem = require('./utils/util-pem');
  const passport = require('passport');
  const morgan = require('morgan');
  const BearerStrategy = require('passport-http-bearer');
  const {ResourceManager} = require('@lgslabs/bits-auto-discovery');
  const SERVER_KEY_SIZE = 4096;

  const logger = require('@lgslabs/bits-logger').getLogger();

  const TOPIC = {
    ROOT_ROUTE: 'base-server-root-router',
    STRATEGY_API: 'passport-strategy-api',
  };

  // Backup files
  const BackUpCrt = path.join(global.paths.data, '/base', '/certs', 'device-server.crt');
  const BackUpKey = path.join(global.paths.data, '/base', '/keys', 'device.pem');

  class BaseServer extends EventEmitter {
    constructor(authManager, keyManager, certManager) {
      super();

      this.authManager = authManager;
      this._keyManager = keyManager;
      this._certManager = certManager;
      this.server = null;
      this._middlewares = {};

      this._routerResourceMgr = new ResourceManager({topic: TOPIC.ROOT_ROUTE});
      this._routerResourceMgr.on('add', this._onRouterAdded.bind(this));

      this._strategyResourceMgr = new ResourceManager({topic: TOPIC.STRATEGY_API});
      this._strategyResourceMgr.on('add', this._onStrategyAdded.bind(this));
      this._strategyResourceMgr.on('remove', this._onStrategyRemoved.bind(this));

      passport.use('bearer', new BearerStrategy(this._validateToken.bind(this)));

      this.app = express();
      this.app.use(helmet());
      // this.app.use(morgan('dev'));
      this.app.use(compression());
      this.app.use(passport.initialize());
    }

    getPassport() {
      return passport;
    }

    _validateToken(token, done) {
      this.authManager.validateAccessToken(token)
      .then((accessToken) => {
        if (accessToken) {
          done(null, accessToken.user, {scope: 'all'});
        } else {
          done(null, false, {message: 'auth/invalid-token'});
        }
      })
      .catch((err) => done(null, false, {message: err.message}));
    }

    _getServerKey(keyManager) {
      return Promise.resolve()
      .then(() => keyManager.getDevicePrivateKey())
      .then((key) => key.getFilepath())
      .then((filename) => UtilFs.readFile(filename, 'utf8'))
      .catch((err) => {
        return UtilPem.createPrivateKey(SERVER_KEY_SIZE)
        .then((key) => {
          return UtilFs.writeFile(BackUpKey, key, 'utf8')
          .then(() => key);
        })
        .then((key) => {
          this._keyManager.reloadKeys();
          return key;
        });
      });
    }

    _getServerCertificate(certManager, key) {
      return Promise.resolve()
      .then(() => certManager.getDeviceServerCert())
      .then((cert) => cert.getFilepath())
      .then((filename) => UtilFs.readFile(filename, 'utf8'))
      .catch((err) => {
        const altNames = [];
        const ifaces = os.networkInterfaces();
        Object.keys(ifaces).forEach((ifacename) => {
          const ifaceInfos = ifaces[ifacename];
          ifaceInfos.forEach((info) => altNames.push(info.address));
        });

        const options = {
          clientKey: key,
          altNames: altNames,
          selfSigned: true,
          days: 999,
        };

        return UtilPem.createCertificate(options)
        .then((certificate) => {
          return UtilFs.writeFile(BackUpCrt, certificate, 'utf8')
          .then(() => certificate);
        })
        .then((certificate) => {
          certManager.reloadCerts();
          return certificate;
        });
      });
    }

    _getServerOptions(keyManager, certManager) {
      const options = {};
      return this._getServerKey(keyManager)
      .then((key) => {
        options.key = key;
        return this._getServerCertificate(certManager, key);
      })
      .then((cert) => {
        options.cert = cert;
        return options;
      });
    }

    load(messageCenter) {
      this._messageCenter = messageCenter;

      return Promise.resolve()
      .then(() => this._routerResourceMgr.load(messageCenter))
      .then(() => this._strategyResourceMgr.load(messageCenter))
      .then(() => this.listen());
    }

    unload() {
      return Promise.resolve()
      .then(() => this._strategyResourceMgr.removeAllListeners())
      .then(() => this._strategyResourceMgr.unload());
    }

    listen() {
      logger.debug('Web Server Listen');
      return Promise.resolve()
      .then(() => this._createHttpServer())
      .then((server) => this._getServerOptions(this._keyManager, this._certManager))
      .then((options) => this._createServer(options));
    }

    _createServer(options) {
      return new Promise((fulfill, reject) => {
        // Create and start the server
        this.server = https.createServer(options, this.app);
        this.server.once('error', reject);
        this.server.listen(HTTPS_PORT, () => fulfill(this.server));
      });
    }

    _createHttpServer() {
      return new Promise((resolve, reject) => {
        const app = express();
        app.use(helmet());
        app.use(function(req, res) {
          res.redirect('https://' + req.hostname + ':' + HTTPS_PORT);
        });

        const server = http.createServer(app);
        server.once('error', reject);
        server.listen(HTTP_PORT, resolve);
      });
    }

    _onStrategyAdded(res) {
      const {name} = res.getValue();
      Promise.resolve()
      .then(() => {
        const StrategyManager = global.AuthStrategies[name];
        const mgr = new StrategyManager();
        return mgr.load(this._messageCenter)
        .then(() => mgr.getInstance());
      })
      .then((instance) => {
        passport.use(name, instance);
        return Promise.resolve(this.app.use(passport.initialize()))
        .then(() => this.emit('base-server-strategy-used', {name}));
      });
    }

    _onStrategyRemoved(res) {
      const {name} = res.getValue();
      Promise.resolve()
      .then(() => passport.unuse(name))
      .then(() => this.emit('base-server-strategy-unused', {name}));
    }

    _onRouterAdded(resource) {
      return Promise.resolve()
      .then(() => {
        const {path, root, type} = resource.getValue();

        switch (type) {
          case 'route':
            try {
              const Api = require(path);
              const api = new Api(this._messageCenter);
              this.app.use(api.middleware);
            } catch (err) {
              logger.err(`Error loading ${TOPIC.ROOT_ROUTE} for ${path}`, err);
            }
            break;
          case 'static':
            this.app.use(express.static(root));
            break;
          default:
            logger.error(`Unknown ${TOPIC.ROOT_ROUTE} type: ${type}`);
            break;
        }
      });
    }

    use(path, middleware) {
      return Promise.resolve()
      .then(() => {
        if (typeof path === 'function') {
          return this.use('/', path);
        }
        if (!Array.isArray(this._middlewares[path])) {
          this._middlewares[path] = [];
          if (path === '/') {
            this.app.use(this._getMiddleware(path));
          } else {
            this.app.use(path, this._getMiddleware(path));
          }
        }
        this._middlewares[path].push(middleware);
        return this._middlewares;
      });
    }

    updateAuthStrategy(path, middleware) {
      return Promise.resolve()
      .then(() => {
        if (this._middlewares[path]) {
          const index = this._middlewares[path].findIndex((item) => 'object' === typeof item && item.hasOwnProperty('auth'));
          if (0 <= index) {
            this._middlewares[path][index] = middleware;
          } else {
            return Promise.reject(new Error('no auth to update'));
          }
        }
        return this._middlewares;
      });
    }

    removeMiddleware(path, middleware) {
      if (typeof path === 'function') {
        return this.removeMiddleware('/', path);
      }
      if (!Array.isArray(this._middlewares[path])) {
        return Promise.reject(new Error('No middleware for path ' + path + '.'));
      }
      const index = this._middlewares[path].indexOf(middleware);
      if (index < 0) {
        return Promise.reject(new Error('No middleware for path ' + path + '.'));
      }
      this._middlewares[path].splice(index, 1);
      return Promise.resolve(this._middlewares);
    }

    _getMiddleware(path) {
      if (path === undefined) {
        return this._getMiddleware('/');
      }
      if (!Array.isArray(this._middlewares[path])) {
        // eslint-disable-next-line prefer-promise-reject-errors
        return Promise.reject(function(req, res, next) { });
      }

      return function(req, res, next) {
        this._walkSubstack(this._middlewares[path], req, res, next);
      }.bind(this);
    }

    _walkSubstack(stack, req, res, next) {
      if (typeof (stack) === 'function') {
        stack = [stack];
      }

      const walkStack = function(i, err) {
        if (err) {
          return next(err);
        }

        if (i >= stack.length) {
          return next();
        }

        let item = stack[i];
        if (item.auth) item = item.func;

        item(req, res, walkStack.bind(null, i + 1));
      };

      walkStack(0);
    }

    static get morgan() {
      return morgan;
    }
  }

  module.exports = BaseServer;
})();
